### PR TITLE
fix(Core/Pets): fix crash at pet resurrect

### LINKED ISSUE #10125: Warlock pets summoned with half life and disappearing on mount-up
### What client do you play on?

enUS

### Faction

- [X] Alliance
- [X] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59
- [ ] 60

### Current Behaviour
Original report: https://github.com/chromiecraft/chromiecraft/issues/2762

Summoning a warlock pet, it is spawning with half life and mana pool.
After mounting and dismounting, the pet is not appearing again.

After multiple relog attempts (i guess the importent difference here is, if you log in mounted or not, with a pet out or not) the pet stays. But, it is always coming up with 50% hp/lifepool, even if healed full before mount-up.

### Expected Blizzlike Behaviour

pet should be summoned with full hp and not be dismissed on mount up.
uppon dismount, it should return as it was when mounting up.

### Source

_No response_

### Steps to reproduce the problem

summon warlock pet -> 50% hp/mana
heal pet
mount up
dismount
pet either missing, or at 50% hp/mana

### Extra Notes

several players confirmed the issue in worldchat
appeared on 2022-01-10 after the updates

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/86f1b87fe60f462a8045ee169b72fe1104140100

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-exchange-npc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### LINKED ISSUE #10167: server crash 2
### Current Behaviour

crash log 1

https://gist.github.com/mpfans/3f0017381a2e1d3976db771962cb5763

crash log 2

https://gist.github.com/mpfans/35353b589b6974e4ec66b086dc2b2de0

### Expected Blizzlike Behaviour

none

### Source

_No response_

### Steps to reproduce the problem

none

### Extra Notes

_No response_

### AC rev. hash/commit

b2df4aa970706442c1748f144b37a913dec1cc9c  + PR 10025

### Operating system

ubuntu 20.04

### Custom changes or Modules

_No response_

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Add new helper functions for pets
- Fix crash at check pet resurrect https://github.com/azerothcore/azerothcore-wotlk/pull/10025#issuecomment-1009166103

## Issues:
- Closes #10167
- Closes #10125

## Tests Performed:
- Locally

## How to Test the Changes:
1. Check pet ressurect
2. `.pet cre`
3. `.npc tame`
4. Check pet hp/power (mana or etc) after mount/dismount

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by temperrr** [comment]
npc tame  and pet cre with pet
image
Without pet
tames fine

pet revive:

if do too quickly

image

Have to wait 15 sec orso before i can res.

**Comment by temperrr** [comment]
mounting up and down removes hp from pet:
image
image

**Comment by temperrr** [comment]
Tell me when you need another test.

**Comment by Winfidonarleyan** [comment]
> Tell me when you need another test.

Okay

**Comment by mpfans** [comment]
update to e6e76f9 + PR 10120,testing

image


image

**Comment by Winfidonarleyan** [comment]
> update to [e6e76f9](https://github.com/azerothcore/azerothcore-wotlk/commit/e6e76f971b4c03ab818912b9e45c08703e9153f1) + PR 10120,testing
> 
> image
> 
> image
> 
> `server restart 10` not work
> 
> only kill the process

you using modules? show `.ser debug`

**Comment by Winfidonarleyan** [comment]
@temperrr can test pet resurrect?

**Comment by Winfidonarleyan** [comment]
@temperrr fixed HP after mount/dismount

**Comment by temperrr** [comment]
tested!

**Review by Nyeriah — APPROVED** [review]

**Review by IntelligentQuantum — APPROVED** [review]

File: src/server/game/Entities/Pet/Pet.cpp [hunk 1]
```diff
@@ -475,6 +458,22 @@ bool Pet::LoadPetFromDB(Player* owner, uint32 petEntry, uint32 petnumber, bool c
...
 
+        if (getPetType() == SUMMON_PET && !current) //all (?) summon pets come with full health when called, but not when they are current
+        {
+            SetPower(POWER_MANA, GetMaxPower(POWER_MANA));
+            SetFullHealth();
+        }
+        else
+        {
+            if (!savedhealth && getPetType() == HUNTER_PET)
+                setDeathState(JUST_DIED);
```
Comment (claudiodfc) [inline src/server/game/Entities/Pet/Pet.cpp:469]: Should check if pet isn't dead, could be possible causing new issue

Comment (Winfidonarleyan) [reply src/server/game/Entities/Pet/Pet.cpp:469]: Okay, thanks

**Pull request merged**